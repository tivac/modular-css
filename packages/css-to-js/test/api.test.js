"use strict";

const path = require("path");

const dedent = require("dedent");
const Processor = require("@modular-css/processor");

const { transform } = require("../css-to-js.js");

const root = path.resolve(__dirname, "../../../").replace(/\\/g, "/");
const safe = "<ROOT-DIR>";

expect.addSnapshotSerializer({
    print(val) {
        return val.replace(root, safe);
    },
    test(val) {
        return typeof val === "string";
    },
});

const resolvers = [
    (src, file) => path.join(path.dirname(src), file),
];

describe("@modular-css/css-to-js API", () => {
    it("should be a function", () =>
        expect(typeof transform).toBe("function")
    );

    it("should generate javascript", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        const { code, namedExports } = transform("./a.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "a" ]);
    });

    it("should generate a javscript proxy in dev", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        const { code } = transform("./a.css", processor, { dev : true });

        expect(code).toMatchSnapshot("code");
    });

    it("should generate empty results & a warning on invalid file input", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        const { code, namedExports, warnings } = transform("./NOPE.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ ]);
        expect(warnings).toMatchSnapshot("warnings");
    });

    it("should represent local composition", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: blue; } .b { composes: a; }`);

        const { code, namedExports } = transform("./a.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "a", "b" ]);
    });

    it("should represent external composition", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        await processor.string("./b.css", `.b { composes: a from "./a.css"; color: blue; }`);

        const { code, namedExports } = transform("./b.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "b" ]);
    });

    it("should use relative imports when requested", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        await processor.string("./b.css", `.b { composes: a from "./a.css"; color: blue; }`);

        const { code, namedExports } = transform("./b.css", processor, { relativeImports : true });

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "b" ]);
    });

    it("should output without default export", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        const { code, namedExports } = transform("./a.css", processor, { defaultExport : false });

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "a" ]);
    });

    it("should output css when requested", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);

        const { code, namedExports } = transform("./a.css", processor, { styleExport : true });

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "a" ]);
    });

    it("should dedupe repeated identifiers", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a { color: red; }`);
        await processor.string("./b.css", `.a { composes: a from "./a.css"; color: blue; }`);

        const { code, namedExports } = transform("./b.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "a1 as a" ]);
    });

    it("should represent local @values", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `@value v: #00F; .a { color: v; }`);

        const { code, namedExports } = transform("./a.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "$values", "a" ]);
    });

    it("should represent external @values", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `@value v: #00F;`);
        await processor.string("./b.css", `@value v from "./a.css"; .b { color: v; }`);

        const { code, namedExports } = transform("./b.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "$values", "b" ]);
    });

    it("should represent external @values namespaces", async () => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `@value v1: #00F; @value v2: #F00; `);
        await processor.string("./b.css", dedent(`
            @value * as values from "./a.css";
            @value v: #0F0;

            .b {
                color: values.v1;
                background-color: values.v2;
                border-color: v;
            }
        `));

        const { code, namedExports } = transform("./b.css", processor);

        expect(code).toMatchSnapshot("code");
        expect(namedExports).toEqual([ "$values", "b" ]);
    });


    it("should generate javascript from composes", async () => {
        const processor = new Processor({ resolvers });

        await processor.file(require.resolve("./specimens/composes/custom-first-rule.css"));

        const { code } = transform(require.resolve("./specimens/composes/custom-first-rule.css"), processor);

        expect(code).toMatchSnapshot("code");
    });

    it("should generate javascript from multiple composes", async () => {
        const processor = new Processor({ resolvers });

        await processor.file(require.resolve("./specimens/composes/custom-between-rules.css"));

        const { code } = transform(require.resolve("./specimens/composes/custom-between-rules.css"), processor);

        expect(code).toMatchSnapshot("code");
    });

    it("should generate javascript with var variable statements", async () => {
        const processor = new Processor({ resolvers });

        await processor.file(require.resolve("./specimens/simple.css"));

        const { code } = transform(require.resolve("./specimens/simple.css"), processor, { variableDeclaration : "var" });

        expect(code).toMatchSnapshot("code");
    });

    it.each([
        true,
        false,
        {},
        { rewriteInvalid : false },
        { warn : false },
        { rewriteInvalid : false, warn : false },
    ])("should handle options.namedExports set to: %s", async (namedExports) => {
        const processor = new Processor({ resolvers });

        await processor.string("./a.css", `.a-1 { color: red; }`);

        const { code, namedExports : exported, warnings } = transform("./a.css", processor, { namedExports });

        expect(code).toMatchSnapshot("code");
        expect(warnings).toMatchSnapshot("warnings");
        expect(exported).toMatchSnapshot("named exports");
    });
});
