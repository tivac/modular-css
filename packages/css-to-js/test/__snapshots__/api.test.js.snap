// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modular-css/css-to-js API should dedupe repeated identifiers: code 1`] = `
import { a } from "<ROOT-DIR>/a.css";
const a1 = a + " " + "mc71966a67_a"
export default {
    "a" : a1
};

export {
    a1 as a
};
`;

exports[`@modular-css/css-to-js API should generate a javscript proxy in dev: code 1`] = `
const a = "mc74f3fa7b_a"
const data = {
    a
};

export default new Proxy(data, {
    get(tgt, key) {
        if(key in tgt) {
            return tgt[key];
        }

        throw new ReferenceError(
            key + " is not exported by " + "a.css"
        );
    }
})

export {
    a
};
`;

exports[`@modular-css/css-to-js API should generate empty results & a warning on invalid file input: code 1`] = `export default null;`;

exports[`@modular-css/css-to-js API should generate empty results & a warning on invalid file input: warnings 1`] = `
Array [
  ./NOPE.css doesn't exist in the processor instance,
]
`;

exports[`@modular-css/css-to-js API should generate javascript from composes: code 1`] = `
const custom = "mca72bf2bd_custom"
const rule = custom + " " + "mca72bf2bd_rule"
export default {
    custom,
rule
};

export {
    custom,
rule
};
`;

exports[`@modular-css/css-to-js API should generate javascript from multiple composes: code 1`] = `
const rule = custom + " " + "mc9c938a76_rule"
const custom = "mc9c938a76_custom"
export default {
    rule,
custom
};

export {
    rule,
custom
};
`;

exports[`@modular-css/css-to-js API should generate javascript with var variable statements: code 1`] = `
var str = "\\"string\\""
var $values = {
    str,
};
var fooga = "mce1ef44f3_fooga"
export default {
    $values,
fooga
};

export {
    $values,
fooga
};
`;

exports[`@modular-css/css-to-js API should generate javascript: code 1`] = `
const a = "mc74f3fa7b_a"
export default {
    a
};

export {
    a
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false }: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false }: named exports 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false }: warnings 1`] = `
Array [
  "a-1" is not a valid JS identifier,
]
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false, warn: false }: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false, warn: false }: named exports 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { rewriteInvalid: false, warn: false }: warnings 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { warn: false }: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};

export {
    a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { warn: false }: named exports 1`] = `
Array [
  a1,
]
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: { warn: false }: warnings 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: {}: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};

export {
    a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: {}: named exports 1`] = `
Array [
  a1,
]
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: {}: warnings 1`] = `
Array [
  "a-1" is not a valid JS identifier, exported as "a1",
]
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: false: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: false: named exports 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: false: warnings 1`] = `Array []`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: true: code 1`] = `
const a1 = "mc74f3fa7b_a-1"
export default {
    "a-1" : a1
};

export {
    a1
};
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: true: named exports 1`] = `
Array [
  a1,
]
`;

exports[`@modular-css/css-to-js API should handle options.namedExports set to: true: warnings 1`] = `
Array [
  "a-1" is not a valid JS identifier, exported as "a1",
]
`;

exports[`@modular-css/css-to-js API should output css when requested: code 1`] = `
const a = "mc74f3fa7b_a"
export default {
    a
};

export {
    a
};
export const styles = ".mc74f3fa7b_a { color: red; }";
`;

exports[`@modular-css/css-to-js API should output without default export: code 1`] = `
const a = "mc74f3fa7b_a"

export {
    a
};
`;

exports[`@modular-css/css-to-js API should represent external @values namespaces: code 1`] = `
import { $values as $aValues } from "<ROOT-DIR>/a.css";
const v = "#0F0"
const $values = {
    "values" : $aValues,
v,
};
const b = "mc71966a67_b"
export default {
    $values,
b
};

export {
    $values,
b
};
`;

exports[`@modular-css/css-to-js API should represent external @values: code 1`] = `
import { $values as $aValues } from "<ROOT-DIR>/a.css";
const $values = {
    "v" : $aValues["v"],
};
const b = "mc71966a67_b"
export default {
    $values,
b
};

export {
    $values,
b
};
`;

exports[`@modular-css/css-to-js API should represent external composition: code 1`] = `
import { a } from "<ROOT-DIR>/a.css";
const b = a + " " + "mc71966a67_b"
export default {
    b
};

export {
    b
};
`;

exports[`@modular-css/css-to-js API should represent local @values: code 1`] = `
const v = "#00F"
const $values = {
    v,
};
const a = "mc74f3fa7b_a"
export default {
    $values,
a
};

export {
    $values,
a
};
`;

exports[`@modular-css/css-to-js API should represent local composition: code 1`] = `
const a = "mc74f3fa7b_a"
const b = a + " " + "mc74f3fa7b_b"
export default {
    a,
b
};

export {
    a,
b
};
`;

exports[`@modular-css/css-to-js API should use relative imports when requested: code 1`] = `
import { a } from "./a.css";
const b = a + " " + "mc71966a67_b"
export default {
    b
};

export {
    b
};
`;
